Add project view command

brodock opened the pull request:
**Description**
Adds `glab project view [repository] [flags]` command.

View, will render project name, description and README. Optionally you can pass `--web` and it will open the project in your default browser instead.

If you don't pass a `repository` path, it will try to infer based on current git repository. If it finds, it will also use current branch to retrieve the README from, or to open the browser project page within that branch.

It also accepts just a repository name without a namespace, in which it will use your own username as the namespace.

Ex: `glab project view glab` will open the `brodock/glab` in my case.

**Related Issue**
Resolves #777 

**How Has This Been Tested?**
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
I've only tested this manually. Will work on adding go tests.

**Screenshots (if appropriate):**

**Types of changes**
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation
- [ ] Chore (Related to CI or Packaging to platforms)

brodock committed:
Allow rendering Markdown without indentations

brodock committed:
Added GetFile api support to allow retrieve file content from repository

brodock committed:
Added project view command.

Usage: glab project view [repository] [--web]

Will either display project name, description and README
or open the URL in the default browser.

If repository is not informed it will try to get from current git
and will use current branch.

profclems commented on commands/project/view/project_view.go:
```diff
@@ -0,0 +1,224 @@
+
+type ViewOptions struct {
+	ProjectID    string
+	ApiClient    *gitlab.Client
```
```suggestion
	APIClient    *gitlab.Client
```

profclems commented on commands/project/view/project_view.go:
```diff
@@ -0,0 +1,224 @@
+		currentUser, err := api.CurrentUser(opts.ApiClient)
+
+		if err != nil {
+			fmt.Fprintf(opts.IO.StdErr, "Failed to retrieve your current user: %s", err)
+
+			return err
```
This will literally prints the error twice. 
The errors are already printed to stderr at the root function so you just have to return the error. Please use `return fmt.Errorf()` if you want to add more info to the error or you can use `cmdutils.WrapError()` for a verbose and pretty printed error message.

```suggestion
return cmdutils.WrapError(err, "Failed to retrieve your current user");
```

profclems commented on commands/project/view/project_view.go:
```diff
@@ -0,0 +1,224 @@
+
+	project, err := api.GetProject(opts.ApiClient, repoPath)
+	if err != nil {
+		fmt.Fprintf(opts.IO.StdErr, "Failed to access project API: %s", err)
+
+		return err
```
Same here

profclems commented on commands/project/view/project_view.go:
```diff
@@ -0,0 +1,224 @@
+	readmeFile, err := api.GetFile(opts.ApiClient, project.PathWithNamespace, readmeFileName, readmeRef)
+
+	if err != nil {
+		fmt.Fprintf(opts.IO.StdErr, "Failed to retrieve README file: %s", err)
+
+		return nil
```
Same here

profclems commented on commands/project/view/project_view.go:
```diff
@@ -0,0 +1,224 @@
+
+	decoded, err := base64.StdEncoding.DecodeString(readmeFile.Content)
+	if err != nil {
+		fmt.Fprintf(opts.IO.StdErr, "Failed to decode README file: %s", err)
+
+		return nil
```
This too

profclems commented on commands/project/view/project_view.go:
```diff
@@ -0,0 +1,224 @@
+			}
+
+			if opts.Branch == "" {
+				opts.Branch, err = git.CurrentBranch()
```
The Factory struct already comes with the current branch in `f.Branch()`. Please use `f.Branch()` instead of `git.CurrentBranch()`

profclems commented on commands/project/view/project_view.go:
```diff
@@ -0,0 +1,224 @@
+
+		return utils.OpenInBrowser(openURL, opts.Browser)
+	} else {
+		if opts.IO.IsaTTY {
```
For TTYs, the output should be piped to the PAGER.

```suggestion
		if err := f.IO.StartPager(); err != nil {
			return err
		}
		defer f.IO.StopPager()

		if opts.IO.IsaTTY {
```

profclems reviewed (changes requested):
Thanks a lot @brodock for working on this! This is great!

Just some few stuff to fix up.

brodock committed:
Use APIClient and f.Branch and handle errors with WrapError

profclems reviewed (approved):
All good! Thanks a lot @brodock

brodock commented:
I've made it into a draft as I found a few edge cases, I will be looking at it tomorrow.
For example, if I type `glab repo view somerepo` while outside a repository (like when in the home directory), it fails because of the first call to `f.Branch()`. So I need to bypass a few things when not in a repository

brodock committed:
Add manual spacing for no description fallback message

This will make the readme be displayed with the same spacing whether
description is present or not.

brodock committed:
Improve spacing for fallback README message

brodock committed:
Allow using `glab view` while outside of a repository

brodock commented:
@profclems this is ready for review again

profclems committed:
add more examples

profclems committed:
add support for full urls

Adding support for allowing users enter full URLs to project path.
This will allows users override the repo host and protocol.

Examples:
- glab repo view git@gitlab.com:profclems/glab.git
- glab repo view https://gitlab.company.org/user/repo
- glab repo view https://gitlab.company.org/user/repo.git

profclems commented:
I have added a few commits to add support for full URLs to allow users specify the repo [git] url as argument.

This will allows users override the repo host and protocol.

Examples:
- `glab repo view git@gitlab.com:profclems/glab.git`
- `glab repo view https://gitlab.company.org/user/repo`
- `glab repo view https://gitlab.company.org/user/repo.git`

brodock commented:
@profclems I've tested the last changes, and it seems it added an unexpected behavior: when I type `glab repo view gitlab-org/gitlab --web` from my current glab branch, it will open `gitlab-org/gitlab` using that branch instead of using default from API.

I will make a new PR to fix this.
